pennylane/templates/subroutines/single_excitation_op.py

```diff
@@ -0,0 +1,160 @@
+    the latter equation is shown below:
+
+    .. figure:: ../../_static/ucc_se_op.png
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    As explained in the paper by Seely *et al.* `arxiv:1208.5986 <https://arxiv.org/abs/1208.5986>`_
+    an :math:`n`-fold tensor product of Pauli-Z matrices requires :math:`2(n-1)` CNOT gates and a
+    single-qubit Z-rotation to exponentiate on a quantum computer. If there are :math:`X` or 
+    :math:`Y` Pauli matrices in the product, the Hadamard (:math:`H`) or :math:`R_x` gate has to 
+    be applied to change to the :math:`X` or :math:`Y` basis, respectively.
+
+    Notice that: 
+
+    #. :math:`\hat{U}_{pr}^{(1)}(\theta)` involves two exponentiations where :math:`\hat{U}_1`,
+       , :math:`\hat{U}_2` and :math:`\hat{U}_\theta` are dfined, for each of them, as follows,
+
+       .. math::
+           [U_1, U_2, U_{\theta})] = \Bigg\{\bigg[R_x(-\pi/2), H, R_z(\theta/2)\bigg], 
+           \bigg[H, R_x(-\frac{\pi}{2}), R_z(-\theta/2) \bigg] \Bigg\}
+
+    #. For a given pair ``[r, p]`` ten single-qubit operations are applied. Notice also that 
+       CNOT gates act only on qubits with indices between ``r`` and ``p``. The operations 
+       performed accross these qubits are shown in dashed lines in the figure above. 
+
+    Args:
+        weight (float): angle :math:`\theta` entering the Z rotation acting on wire ``p``
+        wires (sequence[int]): two-element sequence with the qubit indices ``r, p``
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+    """
+
+    ##############
+    # Input checks
+
+    check_no_variable(wires, msg="'wires' cannot be differentiable")
+
+    wires = check_wires(wires)
+
+    expected_shape = (2,)
+    check_shape(
+        wires,
+        expected_shape,
+        msg="'wires' must be of shape {}; got {}" "".format(expected_shape, get_shape(wires)),
+    )
+
+    expected_shape = ()
+    check_shape(
+        weight,
+        expected_shape,
+        msg="'weight' must be of shape {}; got {}" "".format(expected_shape, get_shape(weight)),
+    )
+
+    check_type(wires, [list], msg="'wires' must be a list; got {}" "".format(wires))
+    for w in wires:
+        check_type(w, [int], msg="'wires' must be a list of integers; got {}" "".format(wires)
+        )
+
+    if wires[1] <= wires[0]:
+        raise ValueError(
+            "wires_1 must be > wires_0; got wires[1]={}, wires[0]={}" ""
+            .format(wires[1], wires[0])
+        )    
+
+    ###############
+
+    r, p = wires
+
+#   Sequence of the wires entering the CNOTs between wires 'r' and 'p'
+    set_cnot_wires = [[l,l+1] for l in range(r,p)]
+
+    n_layers = 2
+    for layer in range(n_layers):
+
+        # U_1, U_2 acting on wires 'r' and 'p'
+        if layer==0:
+            RX(-np.pi/2, wires=r)
+            Hadamard(wires=p)
+        else:
+            Hadamard(wires=r)
+            RX(-np.pi/2, wires=p)
+
```

Comment:
I would have preferred to avoid using conditionals above. I tried using a dictionary e.g.,

```
 s_qubit_gates = {0:{"u1"  : qml.RX(-np.pi/2, wires=r) , "u2"  : qml.Hadamard(wires=p),
                     "rz"  : qml.RZ(weight/2, wires=p) , 
                     "u1a" : qml.RX(np.pi/2, wires=r)  , "u2a" : qml.Hadamard(wires=p)
                     },
                   1:{"u1"  : qml.Hadamard(wires=r)     , "u2"  : qml.RX(-np.pi/2, wires=p), 
                      "rz"  : qml.RZ(-weight/2, wires=p), 
                      "u1a" : qml.Hadamard(wires=r)     , "u2a" : qml.RX(np.pi/2, wires=p)
                   }
 }

for layer in range(n_layers):
    s_qubit_gates[layer]["u1"]
    s_qubit_gates[layer]["u2"]
    ......
```
But this "solution" did not work :-(. The energy coming out of ``qml.VQECost`` was not correct and so was the optimization. Suggestions are welcome. Thanks!.

Reply:
@agran2018, how come you were avoiding using conditionals above? I find the conditionals easier to read and follow than the dictionary approach in your above comment 🙂 

While it might not be our first inclination, one approach could be to make it even more explicit, and not loop over layers (since it looks like the number of layers will always be constant at 2). I think this can have some benefit in terms of code readability:

```python
# --------------------------------------------------
# apply the U1 layer
RX(-np.pi/2, wires=r)
Hadamard(wires=p)

# Applying CNOTs between wires 'r' and 'p'
for cnot_wires in set_cnot_wires:
    CNOT(wires=cnot_wires)

# Z rotation acting on wire 'p'
RZ(weight/2, wires=p)

# Applying CNOTs in reverse order
for cnot_wires in reversed(set_cnot_wires):
    CNOT(wires=cnot_wires)

# U_1^+ acting on wires 'r' and 'p'
RX(np.pi/2, wires=r)
Hadamard(wires=p)


# --------------------------------------------------
# apply the U2 layer
Hadamard(wires=r)
RX(-np.pi/2, wires=p)

# Applying CNOTs between wires 'r' and 'p'
for cnot_wires in set_cnot_wires:
    CNOT(wires=cnot_wires)

# Z rotation acting on wire 'p'
RZ(-weight/2, wires=p)

# Applying CNOTs in reverse order
for cnot_wires in reversed(set_cnot_wires):
    CNOT(wires=cnot_wires)

# U_2^+ acting on wires 'r' and 'p'
Hadamard(wires=r)
RX(np.pi/2, wires=p)
```

Reply:
My guess for why the dict didn't work is that the gates gets evaluated when the dict is constructed and thus being appended to the circuit. I feel it could work if you replace the gates with functions that calls the gates. But I agree with Josh that being explicit here really helps.

Reply:
Yep, that's exactly it @zeyueN.

Reply:
Thanks @josh146 , @zeyueN. I agree. 
Will indicate explicitly ``#apply the first layer`` and ``#apply the second layer`` as U_1 and U_2 notation has to be consistent with the figure.